File: SMO/SmoScripter/VB/SmoScripter/MainForm.vb
'============================================================================
'  File:    MainForm.vb
'
'  Summary: Implements a sample SMO dependency and scripting utility in VB.NET.
'
'  Date:    September 30, 2005
'------------------------------------------------------------------------------
'  This file is part of the Microsoft SQL Server Code Samples.
'
'  Copyright (C) Microsoft Corporation.  All rights reserved.
'
'  This source code is intended only as a supplement to Microsoft
'  Development Tools and/or on-line documentation.  See these other
'  materials for detailed information regarding Microsoft code samples.
'
'  THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
'  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'  PARTICULAR PURPOSE.
'============================================================================

Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Data
Imports System.Collections.Specialized
Imports Microsoft.SqlServer.Management.Smo
Imports Microsoft.SqlServer.Management.Common
Imports Microsoft.SqlServer.MessageBox

''' <summary>
''' Summary description for form.
''' </summary>

Partial Public Class MainForm 'The Partial modifier is only required on one class definition per project.
    Inherits System.Windows.Forms.Form
    ' Use the Server object to connect to a specific server
    Private sqlServerSelection As Server
    Private root As TreeNode

    ''' <summary>
    ''' Required designer variable.
    ''' </summary>
    Private components As System.ComponentModel.IContainer = Nothing

    Public Sub New()
        InitializeComponent()
    End Sub

    Private Sub ConnectToServer(ByVal server As String)
        Try
            Dim srvr As New Server(server)
            srvr.ConnectionContext.ConnectTimeout = 5
            srvr.ConnectionContext.Connect()
            Me.sqlServerSelection = srvr
        Catch ex As ConnectionFailureException
            Dim emb As New ExceptionMessageBox(ex)
            emb.Show(Me)

            Cursor = Cursors.Default
            Me.sqlServerSelection = Nothing
        End Try

        Me.serverVersionToolStripStatusLabel.Text = Me.sqlServerSelection.Information.Version.ToString() + " " + Me.sqlServerSelection.Information.Edition
        Me.SqlServerTreeView.Nodes.Clear()

        ' Add Server node
        Dim node As New TreeNode(Me.sqlServerSelection.ConnectionContext.TrueName)
        node.Name = My.Resources.Server
        root = node
        node.Tag = Me.sqlServerSelection
        Me.SqlServerTreeView.Nodes.Add(node)

        ' Add Databases node
        node = New TreeNode(My.Resources.Databases)
        node.Name = My.Resources.Databases
        node.Tag = Me.sqlServerSelection.Databases
        root.Nodes.Add(node)
        AddDummyNode(node)

        Me.SqlServerTreeView.SelectedNode = root

        ' Optimizing code
        Me.sqlServerSelection.SetDefaultInitFields(GetType(Database), "CreateDate", "IsSystemObject", "CompatibilityLevel")
        Me.sqlServerSelection.SetDefaultInitFields(GetType(Table), "CreateDate", "IsSystemObject")
        Me.sqlServerSelection.SetDefaultInitFields(GetType(Microsoft.SqlServer.Management.Smo.View), "CreateDate", "IsSystemObject")
        Me.sqlServerSelection.SetDefaultInitFields(GetType(StoredProcedure), "CreateDate", "IsSystemObject")
        Me.sqlServerSelection.SetDefaultInitFields(GetType(Column), True)
    End Sub

    Private Sub SqlServerTreeView_AfterSelect(ByVal sender As System.Object, ByVal e As System.Windows.Forms.TreeViewEventArgs) Handles SqlServerTreeView.AfterSelect
        Cursor = Cursors.WaitCursor
        Dim start As DateTime = DateTime.Now
        ShowDetails(Me.SqlServerTreeView.SelectedNode)
        Dim diff As TimeSpan = DateTime.Now - start
        Me.speedToolStripStatusLabel.Text = String.Format(System.Globalization.CultureInfo.InvariantCulture, "{0:####}", diff.TotalMilliseconds)
        Cursor = Cursors.Default
    End Sub

    Private Sub SqlServerTreeView_AfterExpand(ByVal sender As System.Object, ByVal e As System.Windows.Forms.TreeViewEventArgs) Handles SqlServerTreeView.AfterExpand
        Cursor = Cursors.WaitCursor
        LoadTreeViewItems(e.Node)
        Cursor = Cursors.Default
    End Sub

    ' Loads treeview items and updates listview as well.
    <System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2201:DoNotRaiseReservedExceptionTypes")> _
    Private Sub ShowDetails(ByVal node As TreeNode)
        Dim smoObject As SqlSmoObject = Nothing
        Dim smoCollection As SmoCollectionBase = Nothing

        Me.ListView.Items.Clear()

        If node Is Nothing Then
            Return
        End If

        Select Case node.Tag.GetType().Name
            Case "DatabaseCollection", "TableCollection", "ViewCollection", "StoredProcedureCollection", "ColumnCollection", "SqlAssemblyCollection"
                ' Load the items of a collection, if not already loaded
                LoadTreeViewItems(node)

                ' Update the TreeView
                smoCollection = CType(node.Tag, SmoCollectionBase)
                UpdateListViewWithCollection(smoCollection)

            Case "Server"
                smoObject = CType(node.Tag, Server).Information
                UpdateListView(smoObject, True, "Information")

                smoObject = CType(node.Tag, Server).Settings
                UpdateListView(smoObject, False, "Settings")

            Case "Database", "Table", "View", "StoredProcedure", "Column", "SqlAssembly"
                smoObject = CType(node.Tag, SqlSmoObject)
                UpdateListView(smoObject, True, Nothing)

            Case Else
                Throw New Exception(My.Resources.UnrecognizedType _
                    & node.Tag.GetType().ToString())
        End Select
    End Sub

    ' Update the list view for a property list
    Private Sub UpdateListView(ByVal smoObject As SqlSmoObject, ByVal clear As Boolean, ByVal group As String)
        smoObject.Initialize(True)
        If clear = True Then
            Me.ListView.Columns.Clear()
            Me.ListView.Groups.Clear()

            Dim colHeader As New ColumnHeader()
            colHeader.Text = My.Resources.PropertyName
            Me.ListView.Columns.Add(colHeader)

            colHeader = New ColumnHeader()
            colHeader.Text = My.Resources.Value
            Me.ListView.Columns.Add(colHeader)
        End If

        Dim lvGroup As ListViewGroup = Nothing
        If Not (group Is Nothing) Then
            lvGroup = New ListViewGroup(group)
            Me.ListView.Groups.Add(lvGroup)
        End If

        Dim lvi As New ListViewItem()
        lvi.Text = My.Resources.Urn
        lvi.Name = My.Resources.Urn
        lvi.Group = lvGroup
        lvi.SubItems.Add(smoObject.Urn)
        Me.ListView.Items.Add(lvi)

        Dim prop As [Property]
        For Each prop In smoObject.Properties
            If Not (prop.Value Is Nothing) Then
                Dim lvItem As New ListViewItem()
                lvItem.Text = prop.Name
                lvItem.Name = prop.Name
                lvItem.Group = lvGroup

                lvItem.SubItems.Add(prop.Value.ToString())
                Me.ListView.Items.Add(lvItem)
            End If
        Next

        Me.ListView.Sorting = System.Windows.Forms.SortOrder.Ascending
        Me.ListView.Sort()
        Me.ListView.Columns(0).AutoResize(ColumnHeaderAutoResizeStyle.ColumnContent)
        Me.ListView.Columns(1).AutoResize(ColumnHeaderAutoResizeStyle.ColumnContent)
    End Sub

    ' Lists a collection in the listview if the container node is selected
    Private Sub UpdateListViewWithCollection(ByVal smoCollection As SmoCollectionBase)
        Me.ListView.Columns.Clear()
        Me.ListView.Groups.Clear()

        Dim colHeader As New ColumnHeader()
        colHeader.Text = My.Resources.ObjectName
        Me.ListView.Columns.Add(colHeader)

        colHeader = New ColumnHeader()
        If TypeOf smoCollection Is ColumnCollection Then
            colHeader.Text = My.Resources.DataType
        Else
            colHeader.Text = My.Resources.DateCreated
        End If

        Me.ListView.Columns.Add(colHeader)

        For Each smoObject As SqlSmoObject In smoCollection
            If smoObject.Properties.Contains("IsSystemObject") = True _
                AndAlso CBool(smoObject.Properties("IsSystemObject").Value) = True Then
                Continue For
            End If

            Dim lvi As New ListViewItem()
            lvi.Text = smoObject.ToString()
            lvi.Name = smoObject.ToString()

            If TypeOf smoObject Is Column Then
                lvi.SubItems.Add(smoObject.Properties("DataType").Value.ToString())
            Else
                lvi.SubItems.Add(smoObject.Properties("CreateDate").Value.ToString())
            End If

            lvi.Tag = smoObject
            Me.ListView.Items.Add(lvi)
        Next

        Me.ListView.Sorting = System.Windows.Forms.SortOrder.Ascending
        Me.ListView.Sort()
        Me.ListView.Columns(0).AutoResize(ColumnHeaderAutoResizeStyle.ColumnContent)
        Me.ListView.Columns(1).AutoResize(ColumnHeaderAutoResizeStyle.ColumnContent)
    End Sub

    ''' <summary>
    ''' Loads the collection items. This has to be deferred else starting is 
    ''' too expensive.
    ''' </summary>
    ''' <param name="node"></param>
    Private Sub LoadTreeViewItems(ByVal node As TreeNode)
        If node.Nodes.Count = 1 AndAlso Not (node.Nodes(My.Resources.DummyNode) Is Nothing) Then
            node.Nodes(My.Resources.DummyNode).Remove()
        Else
            Return
        End If

        Select Case node.Name
            Case "Databases"
                LoadTreeViewDatabases(node)


            Case "Stored Procedures"
                LoadTreeViewStoredProcedures(node)


            Case "Assemblies"
                LoadTreeViewAssemblies(node)


            Case "Tables"
                LoadTreeViewTables(node)


            Case "Views"
                LoadTreeViewViews(node)


            Case "Columns"
                LoadTreeViewColumns(node)
        End Select
    End Sub

    Private Sub LoadTreeViewDatabases(ByVal node As TreeNode)
        Dim db As Database
        For Each db In CType(node.Tag, DatabaseCollection)
            If db.IsSystemObject = False Then
                Dim dbNode As New TreeNode(db.Name)
                dbNode.Name = db.Name
                dbNode.Tag = db
                root.Nodes("Databases").Nodes.Add(dbNode)
                AddDummyNode(node)

                Dim tcNode As New TreeNode(My.Resources.Tables)
                tcNode.Name = My.Resources.Tables
                tcNode.Tag = db.Tables
                dbNode.Nodes.Add(tcNode)
                AddDummyNode(tcNode)

                Dim vcNode As New TreeNode(My.Resources.Views)
                vcNode.Name = My.Resources.Views
                vcNode.Tag = db.Views
                dbNode.Nodes.Add(vcNode)
                AddDummyNode(vcNode)

                Dim spNode As New TreeNode(My.Resources.StoredProcedures)
                spNode.Name = My.Resources.StoredProcedures
                spNode.Tag = db.StoredProcedures
                dbNode.Nodes.Add(spNode)
                AddDummyNode(spNode)

                If sqlServerSelection.Information.Version.Major >= 9 Then
                    Dim assyNode As New TreeNode(My.Resources.Assemblies)
                    assyNode.Name = My.Resources.Assemblies
                    assyNode.Tag = db.Assemblies
                    dbNode.Nodes.Add(assyNode)
                    AddDummyNode(assyNode)
                End If
            End If
        Next
    End Sub

    Private Shared Sub LoadTreeViewTables(ByVal node As TreeNode)
        Dim tbl As Table
        For Each tbl In CType(node.Tag, TableCollection)
            If tbl.IsSystemObject = False Then
                Dim tNode As New TreeNode(tbl.ToString())
                tNode.Name = tbl.ToString()
                tNode.Tag = tbl
                node.Nodes.Add(tNode)

                ' Add Columns
                Dim tcNode As New TreeNode(My.Resources.Columns)
                tcNode.Name = My.Resources.Columns
                tcNode.Tag = tbl.Columns
                tNode.Nodes.Add(tcNode)
                AddDummyNode(tcNode)
            End If
        Next
    End Sub

    Private Shared Sub LoadTreeViewViews(ByVal node As TreeNode)
        Dim v As Microsoft.SqlServer.Management.Smo.View
        For Each v In CType(node.Tag, ViewCollection)
            If v.IsSystemObject = False Then
                Dim tNode As New TreeNode(v.ToString())
                tNode.Name = v.ToString()
                tNode.Tag = v
                node.Nodes.Add(tNode)

                ' Add Columns
                Dim tcNode As New TreeNode(My.Resources.Columns)
                tcNode.Name = My.Resources.Columns
                tcNode.Tag = v.Columns
                tNode.Nodes.Add(tcNode)
                AddDummyNode(tcNode)
            End If
        Next
    End Sub

    Private Shared Sub LoadTreeViewColumns(ByVal node As TreeNode)
        Dim col As Column
        For Each col In CType(node.Tag, ColumnCollection)
            Dim tNode As New TreeNode(col.ToString())
            tNode.Name = col.ToString()
            tNode.Tag = col
            node.Nodes.Add(tNode)
        Next
    End Sub

    Private Shared Sub LoadTreeViewStoredProcedures(ByVal node As TreeNode)
        Dim sp As StoredProcedure
        For Each sp In CType(node.Tag, StoredProcedureCollection)
            If sp.IsSystemObject = False Then
                Dim tNode As New TreeNode(sp.ToString())
                tNode.Name = sp.ToString()
                tNode.Tag = sp
                node.Nodes.Add(tNode)
            End If
        Next
    End Sub

    Private Shared Sub LoadTreeViewAssemblies(ByVal node As TreeNode)
        Dim assy As SqlAssembly
        For Each assy In CType(node.Tag, SqlAssemblyCollection)
            Dim tNode As New TreeNode(assy.ToString())
            tNode.Name = assy.ToString()
            tNode.Tag = assy
            node.Nodes.Add(tNode)
        Next
    End Sub

    ' Adds a dummy node, so we can expand a node without querying
    ' whether there are any items
    Private Shared Sub AddDummyNode(ByVal node As TreeNode)
        If node.Nodes.Count <> 0 Then
            Return
        End If

        node.Nodes.Add(My.Resources.DummyNode)
        node.Nodes(0).Name = My.Resources.DummyNode
    End Sub

    ' Double click on a listview item causes the tree to be synced.
    Private Sub ListView_DoubleClick(ByVal sender As Object, ByVal e As EventArgs) Handles ListView.DoubleClick
        If ListView.SelectedItems.Count = 0 Then
            Return
        End If

        Dim obj As Object = ListView.SelectedItems(0).Tag
        If TypeOf obj Is SqlSmoObject = False Then
            Return
        End If

        Dim found As TreeNode() = root.Nodes.Find(obj.ToString(), True)

        ' If new objects are added the names may clash
        If found.Length > 1 Then
            Return
        End If

        SqlServerTreeView.SelectedNode = found(0)
        SqlServerTreeView.Focus()
    End Sub

    <System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")> _
    Private Sub MainForm_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
        Dim ServerConn As New ServerConnection()
        Dim scForm As ServerConnect
        Dim dr As DialogResult

        Try
            ' Display the main window first
            Me.Show()
            Application.DoEvents()

            ServerConn = New ServerConnection()
            scForm = New ServerConnect(ServerConn)
            dr = scForm.ShowDialog(Me)
            If dr = Windows.Forms.DialogResult.OK AndAlso ServerConn.SqlConnectionObject.State = ConnectionState.Open Then
                Me.sqlServerSelection = New Server(ServerConn)
                If Not (Me.sqlServerSelection Is Nothing) Then
                    Me.Text = My.Resources.AppTitle + Me.sqlServerSelection.Name

                    ' Refresh database list
                    ConnectToServer(ServerConn.ServerInstance)
                End If
            Else
                Me.Close()
            End If
        Catch ex As Exception
            Dim emb As New ExceptionMessageBox(ex)
            emb.Show(Me)
        End Try
    End Sub

    Private Sub scriptToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs) Handles scriptToolStripMenuItem.Click
        If Me.ListView.SelectedItems.Count = 0 Then
            Return
        End If

        Dim selectedObject As Object = Me.ListView.SelectedItems(0).Tag
        Dim scriptableObject As IScriptable = TryCast(selectedObject, IScriptable)
        If Not (scriptableObject Is Nothing) Then
            Dim strings As StringCollection = scriptableObject.Script()
            Dim frm As New TextOutputForm(strings)
            frm.ShowDialog(Me)
        End If
    End Sub

    Private Sub scriptwithDependenciesToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs) Handles scriptwithDependenciesToolStripMenuItem.Click
        If Me.ListView.SelectedItems.Count = 0 Then
            Return
        End If

        Dim selectedObject As Object = Me.ListView.SelectedItems(0).Tag
        Dim scriptableObject As IScriptable = TryCast(selectedObject, IScriptable)
        If Not (scriptableObject Is Nothing) Then
            Dim frm As New ScriptingForm(Me.sqlServerSelection, CType(selectedObject, SqlSmoObject))
            frm.ShowDialog(Me)
        End If
    End Sub

    Private Sub dependenciesToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs) Handles dependenciesToolStripMenuItem.Click
        If Me.ListView.SelectedItems.Count = 0 Then
            Return
        End If

        Dim selectedObject As Object = Me.ListView.SelectedItems(0).Tag
        Dim scriptableObject As IScriptable = TryCast(selectedObject, IScriptable)
        If Not (scriptableObject Is Nothing) AndAlso Not TypeOf selectedObject Is Database Then
            Dim smoObjects(0) As SqlSmoObject
            smoObjects(0) = CType(selectedObject, SqlSmoObject)

            Dim scripter As New Scripter(Me.sqlServerSelection)
            Dim frm As New DependenciesForm(Me.sqlServerSelection, scripter.DiscoverDependencies(smoObjects, DependencyType.Parents))

            frm.ShowDialog(Me)
        End If
    End Sub

    Private Sub ListViewContextMenuStrip_Opening(ByVal sender As Object, ByVal e As CancelEventArgs) Handles ListViewContextMenuStrip.Opening
        Dim mnuItem As ToolStripMenuItem
        For Each mnuItem In Me.ListViewContextMenuStrip.Items
            mnuItem.Enabled = False
        Next

        If Me.ListView.SelectedItems.Count = 0 Then
            Return
        End If

        Dim smoObject As SqlSmoObject = CType(Me.ListView.SelectedItems(0).Tag, SqlSmoObject)
        Dim scriptableObject As IScriptable = TryCast(smoObject, IScriptable)
        If Not (scriptableObject Is Nothing) Then
            Me.scriptToolStripMenuItem.Enabled = True
        End If

        If Not (scriptableObject Is Nothing) AndAlso Not TypeOf smoObject Is Database Then
            Me.scriptwithDependenciesToolStripMenuItem.Enabled = True
            Me.dependenciesToolStripMenuItem.Enabled = True
        End If
    End Sub
End Class

File: SMO/SmoScripter/CS/SmoScripter/MainForm.cs
/*============================================================================
  File:    MainForm.cs

  Summary: Implements a sample SMO dependency and scripting utility in C#.

  Date:    September 30, 2005
------------------------------------------------------------------------------
  This file is part of the Microsoft SQL Server Code Samples.

  Copyright (C) Microsoft Corporation.  All rights reserved.

  This source code is intended only as a supplement to Microsoft
  Development Tools and/or on-line documentation.  See these other
  materials for detailed information regarding Microsoft code samples.

  THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
  PARTICULAR PURPOSE.
============================================================================*/

using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;
using System.Collections.Specialized;
using Microsoft.SqlServer.Management.Smo;
using Microsoft.SqlServer.Management.Common;
using Microsoft.SqlServer.MessageBox;

namespace Microsoft.Samples.SqlServer
{
    /// <summary>
    /// Summary description for form.
    /// </summary>
    public partial class MainForm : System.Windows.Forms.Form
    {
        // Use the Server object to connect to a specific server
        private Server sqlServerSelection;
        private TreeNode root;

        /// <summary>
        /// Required designer variable.
        /// </summary>
        private System.ComponentModel.IContainer components = null;

        public MainForm()
        {
            InitializeComponent();
        }

        private void ConnectToServer(string server)
        {
            try
            {
                Server srvr = new Server(server);
                srvr.ConnectionContext.ConnectTimeout = 5;
                srvr.ConnectionContext.Connect();
                this.sqlServerSelection = srvr;
            }
            catch (ConnectionFailureException ex)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox(ex);
                emb.Show(this);

                Cursor = Cursors.Default;
                this.sqlServerSelection = null;
            }

            this.serverVersionToolStripStatusLabel.Text
                = this.sqlServerSelection.Information.Version.ToString()
                + " " + this.sqlServerSelection.Information.Edition;
            this.SqlServerTreeView.Nodes.Clear();

            // Add Server node
            TreeNode node = new TreeNode(this.sqlServerSelection.ConnectionContext.TrueName);
            node.Name = Properties.Resources.Server;
            root = node;
            node.Tag = this.sqlServerSelection;
            this.SqlServerTreeView.Nodes.Add(node);

            // Add Databases node
            node = new TreeNode(Properties.Resources.Databases);
            node.Name = Properties.Resources.Databases;
            node.Tag = this.sqlServerSelection.Databases;
            root.Nodes.Add(node);
            AddDummyNode(node);

            this.SqlServerTreeView.SelectedNode = root;

            // Optimizing code
            this.sqlServerSelection.SetDefaultInitFields(typeof(Database),
                "CreateDate", "IsSystemObject", "CompatibilityLevel");
            this.sqlServerSelection.SetDefaultInitFields(typeof(Table),
                "CreateDate", "IsSystemObject");
            this.sqlServerSelection.SetDefaultInitFields(typeof(Microsoft.SqlServer.Management.Smo.View),
                "CreateDate", "IsSystemObject");
            this.sqlServerSelection.SetDefaultInitFields(typeof(StoredProcedure),
                "CreateDate", "IsSystemObject");
            this.sqlServerSelection.SetDefaultInitFields(typeof(Column), true);
        }

        private void SqlServerTreeView_AfterSelect(object sender, TreeViewEventArgs e)
        {
            Cursor = Cursors.WaitCursor;
            DateTime start = DateTime.Now;
            ShowDetails(this.SqlServerTreeView.SelectedNode);
            TimeSpan diff = DateTime.Now - start;
            this.speedToolStripStatusLabel.Text
                = String.Format(System.Globalization.CultureInfo.InvariantCulture,
                "{0:####}", diff.TotalMilliseconds);
            Cursor = Cursors.Default;
        }

        private void SqlServerTreeView_AfterExpand(object sender, TreeViewEventArgs e)
        {
            Cursor = Cursors.WaitCursor;
            LoadTreeViewItems(e.Node);
            Cursor = Cursors.Default;
        }

        // Loads treeview items and updates listview as well.
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2201:DoNotRaiseReservedExceptionTypes")]
        private void ShowDetails(TreeNode node)
        {
            SqlSmoObject smoObject = null;
            SmoCollectionBase smoCollection = null;

            this.ListView.Items.Clear();

            if (node == null)
            {
                return;
            }

            switch (node.Tag.GetType().Name)
            {
                case "DatabaseCollection":
                case "TableCollection":
                case "ViewCollection":
                case "StoredProcedureCollection":
                case "ColumnCollection":
                case "SqlAssemblyCollection":
                    // Load the items of a collection, if not already loaded
                    LoadTreeViewItems(node);

                    // Update the TreeView
                    smoCollection = (SmoCollectionBase)node.Tag;
                    UpdateListViewWithCollection(smoCollection);

                    break;

                case "Server":
                    smoObject = ((Server)node.Tag).Information;
                    UpdateListView(smoObject, true, "Information");

                    smoObject = ((Server)node.Tag).Settings;
                    UpdateListView(smoObject, false, "Settings");

                    break;

                case "Database":
                case "Table":
                case "View":
                case "StoredProcedure":
                case "Column":
                case "SqlAssembly":
                    smoObject = (SqlSmoObject)node.Tag;
                    UpdateListView(smoObject, true, null);

                    break;

                default:
                    throw new Exception(Properties.Resources.UnrecognizedType 
                        + node.Tag.GetType().ToString());
            }
        }

        // Update the list view for a property list
        private void UpdateListView(SqlSmoObject smoObject, bool clear, string group)
        {
            smoObject.Initialize(true);
            if (clear == true)
            {
                this.ListView.Columns.Clear();
                this.ListView.Groups.Clear();

                ColumnHeader colHeader = new ColumnHeader();
                colHeader.Text = Properties.Resources.PropertyName;
                this.ListView.Columns.Add(colHeader);

                colHeader = new ColumnHeader();
                colHeader.Text = Properties.Resources.Value;
                this.ListView.Columns.Add(colHeader);
            }

            ListViewGroup lvGroup = null;
            if (group != null)
            {
                lvGroup = new ListViewGroup(group);
                this.ListView.Groups.Add(lvGroup);
            }

            ListViewItem lvi = new ListViewItem();
            lvi.Text = Properties.Resources.Urn;
            lvi.Name = Properties.Resources.Urn;
            lvi.Group = lvGroup;
            lvi.SubItems.Add(smoObject.Urn);
            this.ListView.Items.Add(lvi);

            foreach (Property prop in smoObject.Properties)
            {
                if (prop.Value != null)
                {
                    ListViewItem lvItem = new ListViewItem();
                    lvItem.Text = prop.Name;
                    lvItem.Name = prop.Name;
                    lvItem.Group = lvGroup;

                    lvItem.SubItems.Add(prop.Value.ToString());
                    this.ListView.Items.Add(lvItem);
                }
            }

            this.ListView.Sorting = System.Windows.Forms.SortOrder.Ascending;
            this.ListView.Sort();
            this.ListView.Columns[0].AutoResize(ColumnHeaderAutoResizeStyle.ColumnContent);
            this.ListView.Columns[1].AutoResize(ColumnHeaderAutoResizeStyle.ColumnContent);
        }

        // Lists a collection in the listview if the container node is selected
        private void UpdateListViewWithCollection(SmoCollectionBase smoCollection)
        {
            this.ListView.Columns.Clear();
            this.ListView.Groups.Clear();

            ColumnHeader colHeader = new ColumnHeader();
            colHeader.Text = Properties.Resources.ObjectName;
            this.ListView.Columns.Add(colHeader);

            colHeader = new ColumnHeader();
            if (smoCollection is ColumnCollection)
            {
                colHeader.Text = Properties.Resources.DataType;
            }
            else
            {
                colHeader.Text = Properties.Resources.DateCreated;
            }

            this.ListView.Columns.Add(colHeader);

            foreach (SqlSmoObject smoObject in smoCollection)
            {
                if (smoObject.Properties.Contains("IsSystemObject") == true
                    && (bool)smoObject.Properties["IsSystemObject"].Value == true)
                {
                    continue;
                }

                ListViewItem lvi = new ListViewItem();
                lvi.Text = smoObject.ToString();
                lvi.Name = smoObject.ToString();

                if (smoObject is Column)
                {
                    lvi.SubItems.Add(smoObject.Properties["DataType"].Value.ToString());
                }
                else
                {
                    lvi.SubItems.Add(smoObject.Properties["CreateDate"].Value.ToString());
                }

                lvi.Tag = smoObject;
                this.ListView.Items.Add(lvi);
            }

            this.ListView.Sorting = System.Windows.Forms.SortOrder.Ascending;
            this.ListView.Sort();
            this.ListView.Columns[0].AutoResize(ColumnHeaderAutoResizeStyle.ColumnContent);
            this.ListView.Columns[1].AutoResize(ColumnHeaderAutoResizeStyle.ColumnContent);
        }

        /// <summary>
        /// Loads the collection items. This has to be deferred else starting is 
        /// too expensive.
        /// </summary>
        /// <param name="node"></param>
        private void LoadTreeViewItems(TreeNode node)
        {
            if (node.Nodes.Count == 1
                && node.Nodes[Properties.Resources.DummyNode] != null)
            {
                node.Nodes[Properties.Resources.DummyNode].Remove();
            }
            else
            {
                return;
            }

            switch (node.Name)
            {
                case "Databases":
                    LoadTreeViewDatabases(node);

                    break;

                case "Stored Procedures":
                    LoadTreeViewStoredProcedures(node);

                    break;

                case "Assemblies":
                    LoadTreeViewAssemblies(node);

                    break;

                case "Tables":
                    LoadTreeViewTables(node);

                    break;

                case "Views":
                    LoadTreeViewViews(node);

                    break;

                case "Columns":
                    LoadTreeViewColumns(node);

                    break;
            }
        }

        private void LoadTreeViewDatabases(TreeNode node)
        {
            foreach (Database db in (DatabaseCollection)node.Tag)
            {
                if (db.IsSystemObject == false)
                {
                    TreeNode dbNode = new TreeNode(db.Name);
                    dbNode.Name = db.Name;
                    dbNode.Tag = db;
                    root.Nodes["Databases"].Nodes.Add(dbNode);
                    AddDummyNode(node);

                    TreeNode tcNode = new TreeNode(Properties.Resources.Tables);
                    tcNode.Name = Properties.Resources.Tables;
                    tcNode.Tag = db.Tables;
                    dbNode.Nodes.Add(tcNode);
                    AddDummyNode(tcNode);

                    TreeNode vcNode = new TreeNode(Properties.Resources.Views);
                    vcNode.Name = Properties.Resources.Views;
                    vcNode.Tag = db.Views;
                    dbNode.Nodes.Add(vcNode);
                    AddDummyNode(vcNode);

                    TreeNode spNode = new TreeNode(Properties.Resources.StoredProcedures);
                    spNode.Name = Properties.Resources.StoredProcedures;
                    spNode.Tag = db.StoredProcedures;
                    dbNode.Nodes.Add(spNode);
                    AddDummyNode(spNode);

                    if (sqlServerSelection.Information.Version.Major >= 9)
                    {
                        TreeNode assyNode = new TreeNode(Properties.Resources.Assemblies);
                        assyNode.Name = Properties.Resources.Assemblies;
                        assyNode.Tag = db.Assemblies;
                        dbNode.Nodes.Add(assyNode);
                        AddDummyNode(assyNode);
                    }
                }
            }
        }

        private static void LoadTreeViewTables(TreeNode node)
        {
            foreach (Table tbl in (TableCollection)node.Tag)
            {
                if (tbl.IsSystemObject == false)
                {
                    TreeNode tNode = new TreeNode(tbl.ToString());
                    tNode.Name = tbl.ToString();
                    tNode.Tag = tbl;
                    node.Nodes.Add(tNode);

                    // Add Columns
                    TreeNode tcNode = new TreeNode(Properties.Resources.Columns);
                    tcNode.Name = Properties.Resources.Columns;
                    tcNode.Tag = tbl.Columns;
                    tNode.Nodes.Add(tcNode);
                    AddDummyNode(tcNode);
                }
            }
        }

        private static void LoadTreeViewViews(TreeNode node)
        {
            foreach (Microsoft.SqlServer.Management.Smo.View v in (ViewCollection)node.Tag)
            {
                if (v.IsSystemObject == false)
                {
                    TreeNode tNode = new TreeNode(v.ToString());
                    tNode.Name = v.ToString();
                    tNode.Tag = v;
                    node.Nodes.Add(tNode);

                    // Add Columns
                    TreeNode tcNode = new TreeNode(Properties.Resources.Columns);
                    tcNode.Name = Properties.Resources.Columns;
                    tcNode.Tag = v.Columns;
                    tNode.Nodes.Add(tcNode);
                    AddDummyNode(tcNode);
                }
            }
        }

        private static void LoadTreeViewColumns(TreeNode node)
        {
            foreach (Column col in (ColumnCollection)node.Tag)
            {
                TreeNode tNode = new TreeNode(col.ToString());
                tNode.Name = col.ToString();
                tNode.Tag = col;
                node.Nodes.Add(tNode);
            }
        }

        private static void LoadTreeViewStoredProcedures(TreeNode node)
        {
            foreach (StoredProcedure sp in (StoredProcedureCollection)node.Tag)
            {
                if (sp.IsSystemObject == false)
                {
                    TreeNode tNode = new TreeNode(sp.ToString());
                    tNode.Name = sp.ToString();
                    tNode.Tag = sp;
                    node.Nodes.Add(tNode);
                }
            }
        }

        private static void LoadTreeViewAssemblies(TreeNode node)
        {
            foreach (SqlAssembly assy in (SqlAssemblyCollection)node.Tag)
            {
                TreeNode tNode = new TreeNode(assy.ToString());
                tNode.Name = assy.ToString();
                tNode.Tag = assy;
                node.Nodes.Add(tNode);
            }
        }

        // Adds a dummy node, so we can expand a node without querying
        // whether there are any items
        private static void AddDummyNode(TreeNode node)
        {
            if (node.Nodes.Count != 0)
            {
                return;
            }

            node.Nodes.Add(Properties.Resources.DummyNode);
            node.Nodes[0].Name = Properties.Resources.DummyNode;
        }

        // Double click on a listview item causes the tree to be synced.
        private void ListView_DoubleClick(object sender, EventArgs e)
        {
            if (ListView.SelectedItems.Count == 0)
            {
                return;
            }

            object obj = ListView.SelectedItems[0].Tag;
            if ((obj is SqlSmoObject) == false)
            {
                return;
            }

            TreeNode[] found = root.Nodes.Find(obj.ToString(), true);

            // If new objects are added the names may clash
            if (found.Length > 1)
            {
                return;
            }

            SqlServerTreeView.SelectedNode = found[0];
            SqlServerTreeView.Focus();
        }

        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
        private void MainForm_Load(object sender, EventArgs e)
        {
            ServerConnection ServerConn = new ServerConnection();
            ServerConnect scForm;
            DialogResult dr;

            try
            {
                // Display the main window first
                this.Show();
                Application.DoEvents();

                ServerConn = new ServerConnection();
                scForm = new ServerConnect(ServerConn);
                dr = scForm.ShowDialog(this);
                if ((dr == DialogResult.OK)
                    && (ServerConn.SqlConnectionObject.State == ConnectionState.Open))
                {
                    this.sqlServerSelection = new Server(ServerConn);
                    if (this.sqlServerSelection != null)
                    {
                        this.Text = Properties.Resources.AppTitle
                            + this.sqlServerSelection.Name;

                        // Refresh database list
                        ConnectToServer(ServerConn.ServerInstance);
                    }
                }
                else
                {
                    this.Close();
                }
            }
            catch (Exception ex)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox(ex);
                emb.Show(this);
            }
        }

        private void scriptToolStripMenuItem_Click(object sender, EventArgs e)
        {
            if (this.ListView.SelectedItems.Count == 0)
            {
                return;
            }

            object selectedObject = this.ListView.SelectedItems[0].Tag;
            IScriptable scriptableObject = selectedObject as IScriptable;
            if (scriptableObject != null)
            {
                StringCollection strings = scriptableObject.Script();
                TextOutputForm frm = new TextOutputForm(strings);
                frm.ShowDialog(this);
            }
        }

        private void scriptwithDependenciesToolStripMenuItem_Click(object sender, EventArgs e)
        {
            if (this.ListView.SelectedItems.Count == 0)
            {
                return;
            }

            object selectedObject = this.ListView.SelectedItems[0].Tag;
            IScriptable scriptableObject = selectedObject as IScriptable;
            if (scriptableObject != null)
            {
                ScriptingForm frm = new ScriptingForm(this.sqlServerSelection,
                    (SqlSmoObject)selectedObject);
                frm.ShowDialog(this);
            }
        }

        private void dependenciesToolStripMenuItem_Click(object sender, EventArgs e)
        {
            if (this.ListView.SelectedItems.Count == 0)
            {
                return;
            }

            object selectedObject = this.ListView.SelectedItems[0].Tag;
            IScriptable scriptableObject = selectedObject as IScriptable;
            if (scriptableObject != null
                && !(selectedObject is Database))
            {
                SqlSmoObject[] smoObjects = new SqlSmoObject[1];
                smoObjects[0] = (SqlSmoObject)selectedObject;

                Scripter scripter = new Scripter(this.sqlServerSelection);
                DependenciesForm frm = new DependenciesForm(this.sqlServerSelection,
                    scripter.DiscoverDependencies(smoObjects, DependencyType.Parents));

                frm.ShowDialog(this);
            }
        }

        private void ListViewContextMenuStrip_Opening(object sender, CancelEventArgs e)
        {
            foreach (ToolStripMenuItem mnuItem in this.ListViewContextMenuStrip.Items)
            {
                mnuItem.Enabled = false;
            }

            if (this.ListView.SelectedItems.Count == 0)
            {
                return;
            }

            SqlSmoObject smoObject = (SqlSmoObject)this.ListView.SelectedItems[0].Tag;
            IScriptable scriptableObject = smoObject as IScriptable;
            if (scriptableObject != null)
            {
                this.scriptToolStripMenuItem.Enabled = true;
            }

            if (scriptableObject != null
                && !(smoObject is Database))
            {
                this.scriptwithDependenciesToolStripMenuItem.Enabled = true;
                this.dependenciesToolStripMenuItem.Enabled = true;
            }
        }
    }

}